klippa/src/serialize.rs

```diff
@@ -188,6 +207,106 @@ impl Serializer {
+        let len;
+        let obj_head;
+        {
+            // code logic is a bit different from Harfbuzz serializer here
+            // because I need to move code around to fix mutable borrow/immutable borrow issue
+            let obj = self.object_pool.get_obj_mut(pool_idx)?;
+
+            self.current = obj.next_obj;
+            obj.tail = self.head;
+            obj.next_obj = None;
+
+            if obj.tail <= obj.head {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return None;
+            }
+
+            len = obj.tail - obj.head;
+
+            // TODO: consider zerocopy
+            // Rewind head
+            self.head = obj.head;
+            self.tail -= len;
+
+            obj_head = obj.head;
+            obj.head = self.tail;
+            obj.tail = self.tail + len;
+
+            //TODO: consider zerocopy here
+            self.data.copy_within(obj_head..obj_head + len, self.tail);
+        }
+
+        let mut hash = 0_u64;
+        // introduce flags to avoid mutable borrow and immutable borrow issues
+        let mut obj_duplicate = false;
+        let mut dup_obj_idxes = (0, 0);
+        if share {
+            hash = hash_one_pool_idx(pool_idx, &self.data, &self.object_pool);
+            if let Some(entry) =
+                self.packed_map
+                    .get_with_hash(pool_idx, hash, &self.data, &self.object_pool)
+            {
+                obj_duplicate = true;
+                dup_obj_idxes = *entry;
+            };
+        }
+
+        if obj_duplicate {
+            self.merge_virtual_links(pool_idx, dup_obj_idxes);
+            let obj = self.object_pool.get_obj_mut(pool_idx).unwrap();
+            obj.fini();
+            self.object_pool.release(pool_idx);
+
+            // rewind tail because we discarded duplicate obj
+            self.tail += len;
+            return Some(dup_obj_idxes.1);
+        }
+
+        self.packed.push(Some(pool_idx));
+        let obj_idx = self.packed.len() - 1;
+
+        if share {
+            self.packed_map
+                .set_with_hash(pool_idx, hash, obj_idx, &self.data, &self.object_pool);
+        }
+        Some(obj_idx)
+    }
+
+    fn merge_virtual_links(&mut self, from: PoolIdx, to: (PoolIdx, ObjIdx)) {
+        let from_obj = self.object_pool.get_obj_mut(from).unwrap();
+        if from_obj.real_links.is_empty() {
+            return;
+        }
+        let mut from_vec = mem::take(&mut from_obj.real_links);
+
+        {
+            // hash value will change after merge, so delete existing old entry in the hash table
+            let hash_old = hash_one_pool_idx(to.0, &self.data, &self.object_pool);
+            self.packed_map
+                .del(to.0, hash_old, &self.data, &self.object_pool);
+        }
+
+        {
+            let to_obj = self.object_pool.get_obj_mut(to.0).unwrap();
+            to_obj.virtual_links.append(&mut from_vec);
+        }
+
```

Comment:
I don't think these braces are doing anything? (this still compiles fine without them)

Reply:
Oh yeah it seems these braces are not needed with mem::take. Thank you!